Mathlib/RingTheory/NonUnitalSubring/Basic.lean

```diff
@@ -0,0 +1,1045 @@
+
+* `Prod A B : NonUnitalSubring (R × S)` : the product of non-unital subrings
+
+* `f.range : NonUnitalSubring B` : the range of the non-unital ring homomorphism `f`.
+
+* `eq_locus f g : NonUnitalSubring R` : given non-unital ring homomorphisms `f g : R →ₙ+* S`,
+     the non-unital subring of `R` where `f x = g x`
+
+## Implementation notes
+
+A non-unital subring is implemented as a non-unital `NonUnitalSubsemiring` which is also an
+additive subgroup.
+
+Lattice inclusion (e.g. `≤` and `⊓`) is used rather than set notation (`⊆` and `∩`), although
+`∈` is defined as membership of a non-unital subring's underlying set.
+
+## Tags
+non-unital subring
+-/
+
+
+open scoped BigOperators
+
+universe u v w
+
+section Basic
+
+variable {R : Type u} {S : Type v} {T : Type w} [NonUnitalNonAssocRing R]
+
+section NonUnitalSubringClass
+
+/-- `NonUnitalSubringClass S R` states that `S` is a type of subsets `s ⊆ R` that
+are both a multiplicative submonoid and an additive subgroup. -/
+class NonUnitalSubringClass (S : Type _) (R : Type u) [NonUnitalNonAssocRing R]
+    [SetLike S R] extends NonUnitalSubsemiringClass S R, NegMemClass S R : Prop where
+
+-- See note [lower instance priority]
+instance (priority := 100) NonUnitalSubringClass.addSubgroupClass (S : Type _) (R : Type u)
+    [SetLike S R] [NonUnitalNonAssocRing R] [h : NonUnitalSubringClass S R] :
+    AddSubgroupClass S R :=
+  { h with }
+
+variable [SetLike S R] [hSR : NonUnitalSubringClass S R] (s : S)
+
+namespace NonUnitalSubringClass
+
+-- Prefer subclasses of `NonUnitalRing` over subclasses of `NonUnitalSubringClass`.
+/-- A non-unital subring of a non-unital ring inherits a non-unital ring structure -/
+instance (priority := 75) toNonUnitalNonAssocRing : NonUnitalNonAssocRing s :=
+  Subtype.val_injective.nonUnitalNonAssocRing _ rfl (fun _ _ => rfl) (fun _ _ => rfl) (fun _ => rfl)
+    (fun _ _ => rfl) (fun _ _ => rfl) fun _ _ => rfl
+
+-- Prefer subclasses of `NonUnitalRing` over subclasses of `NonUnitalSubringClass`.
+/-- A non-unital subring of a non-unital ring inherits a non-unital ring structure -/
+instance (priority := 75) toNonUnitalRing {R : Type _} [NonUnitalRing R] [SetLike S R]
+    [NonUnitalSubringClass S R] (s : S) : NonUnitalRing s :=
+  Subtype.val_injective.nonUnitalRing _ rfl (fun _ _ => rfl) (fun _ _ => rfl) (fun _ => rfl)
+    (fun _ _ => rfl) (fun _ _ => rfl) fun _ _ => rfl
+
+-- Prefer subclasses of `NonUnitalRing` over subclasses of `NonUnitalSubringClass`.
+/-- A non-unital subring of a `NonUnitalCommRing` is a `NonUnitalCommRing`. -/
+instance (priority := 75) toNonUnitalCommRing {R} [NonUnitalCommRing R] [SetLike S R]
+    [NonUnitalSubringClass S R] : NonUnitalCommRing s :=
+  Subtype.val_injective.nonUnitalCommRing _ rfl (fun _ _ => rfl) (fun _ _ => rfl) (fun _ => rfl)
+    (fun _ _ => rfl) (fun _ _ => rfl) fun _ _ => rfl
+
+/-- The natural non-unital ring hom from a non-unital subring of a non-unital ring `R` to `R`. -/
+def subtype (s : S) : s →ₙ+* R :=
```

Comment:
I think it's unfortunate that we tend to use this CS terminology instead of the mathematician's:
```suggestion
def incl (s : S) : s →ₙ+* R :=
```
I guess you should probably stay consistent with `NonUnitalSubsemiringClass.subtype` (and many other examples) but if someone did want rename all these `def subtype`s into `def incl`s then I would merge such a PR!

Reply:
while I agree with the sentiment, we already have things like [docs#Subalgebra.inclusion](https://leanprover-community.github.io/mathlib4_docs/find/?pattern=Subalgebra.inclusion#doc) for the identity inclusion (i.e., `Subtype.map id`) between subalgebras, so I think the subtype terminology for the inclusion into the full type (as opposed to `⊤`) makes sense.